frame/dvm/ethereum/src/mock.rs

```diff
@@ -239,6 +240,122 @@ impl darwinia_ethereum::Config for Test {
 	type StateRoot = IntermediateStateRoot;
 }
 
+// --- pallet-bridge-dispatch config start ---
+pub(crate) type BridgeMessageId = [u8; 4];
+pub(crate) type SubChainId = [u8; 4];
+pub(crate) const SOURCE_CHAIN_ID: SubChainId = *b"srce";
+pub(crate) const TARGET_CHAIN_ID: SubChainId = *b"trgt";
+
+pub struct AccountIdConverter;
+impl sp_runtime::traits::Convert<H256, AccountId32> for AccountIdConverter {
+	fn convert(hash: H256) -> AccountId32 {
+		AccountId32::new(hash.0)
+	}
+}
+
+#[derive(Decode, Encode)]
+pub struct EncodedCall(pub Vec<u8>);
+impl From<EncodedCall> for Result<Call, ()> {
+	fn from(call: EncodedCall) -> Result<Call, ()> {
+		Call::decode(&mut &call.0[..]).map_err(drop)
+	}
+}
+
+pub struct CallValidator;
+impl CallValidate<AccountId32, Origin, Call> for CallValidator {
+	fn pre_dispatch(
+		relayer_account: &AccountId32,
+		origin: &Origin,
+		call: &Call,
+	) -> Result<(), TransactionValidityError> {
+		match call {
+			// Note: Only supprt Ethereum::transact(LegacyTransaction)
+			Call::Ethereum(crate::Call::transact { transaction: tx }) => {
+				match origin.caller() {
+					OriginCaller::Ethereum(RawOrigin::EthereumTransaction(id)) => match tx {
+						Transaction::Legacy(t) => {
+							// Only non-payable call supported.
+							if t.value != U256::zero() {
+								return Err(TransactionValidityError::Invalid(
+									InvalidTransaction::Payment,
+								));
+							}
+							let fee = t.gas_limit.saturating_mul(t.gas_limit);
+
+							let derived_substrate_address = <Test as darwinia_evm::Config>::IntoAccountId::derive_substrate_address(*id);
+							if <Test as darwinia_evm::Config>::RingAccountBasic::account_balance(
+								relayer_account,
+							) >= fee
+							{
+								// Ensure the derived ethereum address has enough balance to pay for
+								// the transaction
+								let _ = <Test as darwinia_evm::Config>::RingAccountBasic::transfer(
+									&relayer_account,
+									&derived_substrate_address,
+									fee,
+								);
+							}
+							let res = Ethereum::validate_transaction_in_block(*id, tx);
```

Comment:
Because this validation is not transactional, and it is too late to invalid here because transfer already happened, or balance storage has already changed. Still worried the attack to run out relayer's balance

So we should still do add some check's before transfer:
`gas_price * gas_limit / 10 ** 9 < MAX_REQUIRED_RELAYER_BALNCE`

Since the settlement of fee market will give the dispatch fees(gas) to relayer regardless the result(true/false) or the reason(failed in validation/execution, e.g. nonce issue), so the transfer is good to happen before `validate_transaction_in_block `: relayer just forward the gas to remote account controlled by message sender.